File: iosApp/iosApp/Widget/PlayerDisplay.swift
//
//  PlayerDisplay.swift
//  iosApp
//
//  Created by Chukwuemeka Nwagu on 05/07/2021.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import sharedmodels

struct PlayerDisplay: View {
    var playViewModel: PlayViewModel
    var color: ChessPieceColor
    
    @ObservedObject
    var boardChanged: Collector<Int32>
    
    init(playViewModel: PlayViewModel, color: ChessPieceColor) {
        self.playViewModel = playViewModel
        self.color = color
        boardChanged = playViewModel.boardUpdated.collectAsObservable(initialValue: playViewModel.boardUpdated.value as! Int32)
    }
    
    var body: some View {
        let player = (color == ChessPieceColor.black)
            ? playViewModel.game.blackPlayer as! GUIPlayer
            : playViewModel.game.whitePlayer as! GUIPlayer
        let isTurn = playViewModel.game.board.turn == color
        HStack {
            ZStack(alignment: .topTrailing) {
                Image(player.avatar)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
                    .frame(width: 25, height: 25)
                if (isTurn) {
                    Circle().fill(Color.green).frame(width:12)
                }
            }
            if (isTurn) {
                Text(playViewModel.game.board.checkMessageForNextPlayer())
                    .foregroundColor(Color.red)
            }
            Spacer()
        }
        .onTapGesture {
            playViewModel.getNextMove()
        }
    }
}

struct PlayerDisplay_Previews: PreviewProvider {
    static var previews: some View {
        PlayerDisplay(playViewModel: PlayViewModel(), color: .white)
    }
}

File: androidApp/src/main/java/com/nwagu/chessboy/widgets/PlayerDisplay.kt
package com.nwagu.chessboy.widgets

import androidx.compose.animation.ExperimentalAnimationApi
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import com.nwagu.chess.gamelogic.turn
import com.nwagu.chess.model.ChessPieceColor
import com.nwagu.chessboy.sharedmodels.players.GUIPlayer
import com.nwagu.chessboy.sharedmodels.presentation.PlayViewModel
import com.nwagu.chessboy.sharedmodels.utils.checkMessageForNextPlayer

@ExperimentalAnimationApi
@Composable
fun PlayerDisplay(
    modifier: Modifier = Modifier,
    viewModel: PlayViewModel,
    color: ChessPieceColor
) {

    val boardChanged by viewModel.boardUpdated.collectAsState(0)

    val isTurn = viewModel.game.board.turn == color

    val player =
        if (color == ChessPieceColor.BLACK) viewModel.game.blackPlayer
        else viewModel.game.whitePlayer

    player as GUIPlayer

    Row(modifier = modifier
        .height(64.dp)
        .padding(16.dp, 8.dp)
        .clickable(onClick = { viewModel.getNextMove() }),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween
    ) {

        // Invisible text to force recompose on board changed
        Text(text = boardChanged.toString(), Modifier.size(0.dp))

        Box {
            Image(
                modifier = Modifier
                    .align(Alignment.Center)
                    .width(25.dp)
                    .aspectRatio(1f),
                painter = painterResource(id = player.avatar),
                contentDescription = "Player avatar",
                contentScale = ContentScale.Fit
            )
            if (isTurn) {
                Box(
                    modifier = modifier
                        .align(Alignment.BottomEnd)
                        .width(12.dp)
                        .aspectRatio(1f)
                        .clip(CircleShape)
                        .background(Color.Green)
                )
            }
        }

        if (isTurn) {
            Text(
                modifier = Modifier,
                text = viewModel.game.board.checkMessageForNextPlayer(),
                color = Color.Magenta
            )
        }

    }
}